datastream/tools/split_dataframe.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+from typing import Tuple, Union, Dict, Optional
+from pathlib import Path
+import numpy as np
+import pandas as pd
+
+from datastream import json
+
+
+def split_dataframe(
+    dataframe: pd.DataFrame,
+    key_column: str,
+    proportions: Dict[str, float],
+    filepath: Optional[Path] = None,
+    stratify_column: Optional[str] = None,
+):
+    '''
+    Split and save result. Add new examples and continue from the old split.
+
+    As new examples come in it can handle:
+    - Changing test size
+    - Adapt after removing examples from dataset
+    - Adapt to new stratification
+    '''
+    if abs(sum(proportions.values()) - 1.0) >= 1e-6:
+        raise AssertionError('Expected sum of proportions to be 1')
+    
+    if filepath is not None and filepath.exists():
+        split = json.read(filepath)
+
+        if set(proportions.keys()) != set(split.keys()):
+            raise AssertionError(
```

Comment:
https://docs.python.org/3/library/exceptions.html#ValueError

Reply:
To add to this we might want to also print what values were received.

Reply:
> It's not immediately obvious to the user how to use `Dataset.split` on a dataset created with `Dataset.from_subscriptable`. Do we want to support this? Different proposals:
> 
> * Change `Dataset.from_subscriptable` dataframe key from `'example'` to `'key'`, have default key name in `Dataset.split` be `'key'`
> * Let the user input (optionally) a column name for the dataframe created by `Dataset.from_subscriptable` (I prefer this)

We do not support Dataset.from_subscriptable for this. We might want to make a note of that. Dataset.subset is also not available for datasets created with from_subscriptable.

I prefer reducing the number of ways to input data over trying to support multiple types of input. It becomes messy and buggy very quickly. I'd rather drop Dataset.from_subscriptable and have a clean interface over trying to support many types of input arguments in all our functions.

_Note that Dataset.from_subscriptable basically does nothing. It just wraps the input list in a pandas.DataFrame_

> Create ``Dataset`` based on subscriptable i.e. implements ``__get_item__``. Mostly useful for simple examples.

In the Dataset.from_subscriptable we say that it's mostly useful for simple examples but we could be more explicit. I'll note these limitations so it's known until we might make a bigger decision.